packages/core/src/stateUtils.ts

```diff
@@ -1358,23 +1357,73 @@ export function microstep<TContext, TEvent extends EventObject>(
 
   actions.push(...res.actions);
 
+  const {
+    actions: resolvedActions,
+    raised,
+    context
+  } = resolveActionsAndContext(actions, machine, _event, currentState);
+
+  internalQueue.push(...res.internalQueue);
+  internalQueue.push(...raised.map(a => a._event));
+
   return {
-    actions,
+    actions: resolvedActions,
     configuration: mutConfiguration,
-    historyValue
+    historyValue,
+    internalQueue,
+    context
   };
 }
 
-export function resolveTransition<
+function selectEventlessTransitions<TContext, TEvent extends EventObject>(
```

Comment:
this new algorithm means that we probably can remove `isTransient` property from StateNode